command/ca/policy/actions/dns.go

```diff
@@ -0,0 +1,144 @@
+package actions
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/urfave/cli"
+
+	"github.com/smallstep/cli/command/ca/policy/policycontext"
+	"github.com/smallstep/cli/flags"
+	"github.com/smallstep/cli/internal/command"
+	"github.com/smallstep/cli/utils/cautils"
+)
+
+var provisionerFilterFlag = cli.StringFlag{
+	Name:  "provisioner",
+	Usage: `The provisioner <name>`,
+}
+
+// DNSCommand returns the dns policy subcommand.
+func DNSCommand(ctx context.Context) cli.Command {
+	return cli.Command{
+		Name:  "dns",
+		Usage: "...",
+		UsageText: `**dns** <domain> [**--remove**]
+[**--provisioner**=<name>] [**--key-id**=<key-id>] [**--reference**=<reference>]
+[**--admin-cert**=<file>] [**--admin-key**=<file>]
+[**--admin-provisioner**=<string>] [**--admin-subject**=<string>]
+[**--password-file**=<file>] [**--ca-url**=<uri>] [**--root**=<file>]
+[**--context**=<name>]`,
+		Description: `**dns** command group provides facilities for ...`,
+		Action: command.InjectContext(
+			ctx,
+			dnsAction,
+		),
+		Flags: []cli.Flag{
+			provisionerFilterFlag,
+			flags.KeyID,
+			flags.Reference,
+			flags.AdminCert,
+			flags.AdminKey,
+			flags.AdminProvisioner,
+			flags.AdminSubject,
+			flags.PasswordFile,
+			flags.CaURL,
+			flags.Root,
+			flags.Context,
+			cli.BoolFlag{
+				Name:  "remove",
+				Usage: `removes the provided DNS names from the policy instead of adding them`,
+			},
+		},
+	}
+}
+
+func dnsAction(ctx context.Context) (err error) {
+
+	clictx := command.CLIContextFromContext(ctx)
+
+	args := clictx.Args()
+	if len(args) == 0 {
+		return errors.New("please provide at least one domain")
+	}
+
+	client, err := cautils.NewAdminClient(clictx)
+	if err != nil {
+		return fmt.Errorf("error creating admin client: %w", err)
+	}
+
+	policy, err := retrieveAndInitializePolicy(ctx, client)
+	if err != nil {
+		return fmt.Errorf("error retrieving policy: %w", err)
+	}
+
+	var dns []string
+
+	switch {
+	case policycontext.HasSSHHostPolicy(ctx):
+		switch {
+		case policycontext.HasAllow(ctx):
+			dns = policy.Ssh.Host.Allow.Dns
+		case policycontext.HasDeny(ctx):
+			dns = policy.Ssh.Host.Deny.Dns
+		default:
+			panic(errors.New("no allow nor deny context set"))
+		}
+	case policycontext.HasSSHUserPolicy(ctx):
+		return errors.New("SSH user policy does not support DNS names")
+	case policycontext.HasX509Policy(ctx):
+		switch {
+		case policycontext.HasAllow(ctx):
+			dns = policy.X509.Allow.Dns
+		case policycontext.HasDeny(ctx):
+			dns = policy.X509.Deny.Dns
+		default:
+			panic(errors.New("no allow nor deny context set"))
+		}
+	default:
+		panic("no SSH nor X.509 context set")
+	}
+
+	if clictx.Bool("remove") {
+		for _, domain := range args {
+			dns = remove(domain, dns)
+		}
+	} else {
+		dns = append(dns, args...)
+	}
```

Comment:
We should try international names with SSH hosts certificates

Reply:
I agree with you that it's good to ensure they're always properly internationalized. However, I think we shouldn't enforce this in the CLI, but should always validate it in (just) the API. The CLI will always call the API, so IMO there's no need to duplicate the logic for these checks. This way control over the rules will be easier, since we don't have to ensure backwards compatibility between the CLI and CA, since there's only one doing the checks. This will also work for other clients interacting with the API.

As for saving the contents in the DB internationalized: that is an option, but I think it's more user friendly to always show the non-internalized names, so that they're more easily readable. If they're valid UTF8 they should end up in the DB OK. Whenever they're requested from the CLI (or a different system), they won't have to be transformed.